### ISSUE TITLE
Merge custom links with auto links.

### ISSUE THREAD

**Issue by lukasoppermann** [thephpleague/fractal#264]
Hey, so I am using fractal with the json api serializier which automatically adds the self link. I want to add another link to a sub-resource, but this is always put into the attributes, not merged with the links.

My transformer returns this:

``` php
return [
            'id'    => $collection->id,
            'name' => $collection->name,
            'slug' => $collection->slug,
            'links' => [[
               'rel' => 'posts',
               'uri' => '/collections/'.$collection->id.'/posts',
            ]]
        ];
```

Using this does not work either:

``` php
return [
            'id'    => $collection->id,
            'name' => $collection->name,
            'slug' => $collection->slug,
            'links' => [
               'posts' => $_ENV['API_DOMAIN'].'/collections/'.$collection->id.'/posts',
            ]
        ];
```

How can I achieve this?

**Comment by lukasoppermann** [comment]
@willishq Any news in this?

**Comment by willishq** [comment]
First, sorry for not getting back to you quickly, I've been pretty hectic.

I'm just reading through the JSON API serializer and spec to see what the best way to proceed with this would be, I'm not seeing a way in either to add links as so, the preferred method appears to be via relationships, @felixkiss might be better to ask specifically about this if they have the time.

**Comment by lukasoppermann** [comment]
Hey @willishq thanks for getting back to me. I think the code I wrote above is not very sensible.

Best solution would probably be:

**1. automatically add all resources that are define in availableIncludes as relationships**

```php
// PostTransformer.php
$availableIncludes = [
  'images',
  'comments',
];
```

should result in the following relationships on `posts` (which as far as I understand is the correct way for json api):

```javascript
{"data": [
    {
        "type": "post",
        "id": "74023261-c072-368f-9ae7-722057446ac6",
        "attributes": {},
        "links": {
          "self": "http://api.formandsystem.app/pages/74078061-c072-368f-9ae7-722057446ac6"
        },
        "relationships": {
            "images": {
                "links": {
                    "self": "http://api.app/posts/74078061-c072-368f-9ae7-722057446ac6/relationships/images",
                    "related": "http://api.app/posts/74078061-c072-368f-9ae7-722057446ac6/images"
                }
            },
            "comments": {
                "links": {
                    "self": "http://api.app/posts/74078061-c072-368f-9ae7-722057446ac6/relationships/comments",
                    "related": "http://api.app/posts/74078061-c072-368f-9ae7-722057446ac6/comments"
                }
            }
        }
    },
]}
```
**2. Links that are specified in the `link` part in a transformer should be merge.**
For example one might want to return a link to the next and previous post. This could be achieved similar to how the `id` is currently transfered from the `attributes` part to the `id` field outside of attributes in the `JsonApiSerializer`.

```php
// PostTransformer.php
transform(Post $post){
        return [
            'id'    => $post->id,
            'links' => [
                'next' => $_ENV['base_url'].'/posts/'.$post->next->id, // pseudo code
                'previous' => $_ENV['base_url'].'/posts/'.$post->prev->id,  // pseudo code
            ]
        ];
}
```

I could send a PR for part 2, if you want. I looked into 1, but could not really figure it out.

**Comment by jonaholsson** [comment]
+1 for above

**Comment by lukasoppermann** [comment]
Before #279 should be merge though. @willishq any idea when you will get around merging it?

**Comment by willishq** [comment]
I just did! I'm gonna fix merge conflicts in your pull request, unless you want to fix them yourself and I'll get it merged in for now!

**Comment by lukasoppermann** [comment]
@willishq awesome, I will try to send a PR as soon as I can.

**Comment by lukasoppermann** [comment]
This will be partly fixed in #284 (fixed 2. Links that are specified in the link part in a transformer should be merge.)
